File: Assistant/lib/utility/wrapper.dart
import '/common.dart'; // ignore: unused_import

// ----------------

class Wrapper<TValue> {

  // #region constructor

  _WrapperData _data;

  // ----------------

  Wrapper(
  ) :
    this._data = .new(null);

  Wrapper.of(
    TValue value,
  ) :
    this._data = .new(value);

  Wrapper._of(
    _WrapperData data,
  ) :
    this._data = data;

  // #endregion

  // #region query

  TValue get value => this._data.value as TValue;
  set value(TValue value) => this._data.value = value;

  // ----------------

  Wrapper<TTarget> cast<TTarget>(
  ) {
    return ._of(this._data);
  }

  // #endregion

}

class _WrapperData {
  Object? value;
  _WrapperData(
    this.value,
  );
}

File: AssistantPlus/Application/Utility/Wrapper.cs
#pragma warning disable 0,
// ReSharper disable

using AssistantPlus;

namespace AssistantPlus.Utility {

	public class Wrapper<TValue>
		where TValue : notnull {

		#region structor

		private WrapperData mData;

		// ----------------

		public Wrapper (
		) {
			this.mData = new ();
			return;
		}

		public Wrapper (
			TValue? value
		) {
			this.mData = new () { Value = value };
			return;
		}

		private Wrapper (
			WrapperData data
		) {
			this.mData = data;
			return;
		}

		#endregion

		#region query

		public TValue? Value {
			get => (TValue?)this.mData.Value;
			set => this.mData.Value = value;
		}

		// ----------------

		public Wrapper<TTarget> Cast<TTarget> (
		)
			where TTarget : class {
			return new (this.mData);
		}

		#endregion

	}

	public record WrapperData {
		public Object? Value = default!;
	}

}
